src/main/java/com/fasterxml/jackson/databind/util/StdDateFormat.java

```diff
@@ -430,161 +388,308 @@ protected boolean looksLikeISO8601(String dateStr)
+        		// truncate if more than 3 digits
+        		if( millisLength > 3 ) {
+        			sb.delete(millisStartPos+4, current);
+        		}
+        		if( millisLength < 3 ) {
+        			sb.insert(current, "0");
+        		}
+        		if( millisLength < 2 ) {
+        			sb.insert(current, "0");
+        		}
+    		}
+    		
+    		current = millisStartPos;
+    	}
+    	
+    	
+    	//
+    	// -- TIME fields --
+    	// 
+    	//  Things to handle:
+    	//  - hours, minutes and seconds are optional. Add a default value so it can be 
+    	//    accepted by SimpleDateFormat
+    	
+    	// Count how many ':' we have in the time part
+    	int columnCount = 0;
+    	for( int i=current-1; i > dateTimeSeparatorPos; i--) {
+    		char cc = dateStr.charAt(i);
+    		if( cc == ':' ) {
+    			columnCount++;
+    		}
+    	}
+    	
+    	if( columnCount < 2 ) { 
+    		sb.insert(current, ":00");
+    	}
+    	if( columnCount < 1 ) {
+    		sb.insert(current, ":00");
+    	}
+	
+        	
+        // Do the actual parsing
+        //
+    	return doParse(sb.toString(), df);
     }
 
-    private final static boolean hasTimeZone(String str)
+    protected Date doParse(final String dateStr, final SimpleDateFormat baselineDateFormat) throws ParseException 
     {
-        // Only accept "+hh", "+hhmm" and "+hh:mm" (and with minus), so
-        int len = str.length();
-        if (len >= 6) {
-            char c = str.charAt(len-6);
-            if (c == '+' || c == '-') return true;
-            c = str.charAt(len-5);
-            if (c == '+' || c == '-') return true;
-            c = str.charAt(len-3);
-            if (c == '+' || c == '-') return true;
+    	// Clone the baseline format given as input so as to reuse any pre-configured 
+    	// instance cached locally.
+    	DateFormat df = _cloneFormatIfNeeded(baselineDateFormat);
+    	
+    	// Do the actual parsing
+    	//
+        Date dt = df.parse(dateStr);
+        
+        // 22-Dec-2015, tatu: With non-lenient, may get null
+        if (dt == null) {
+            throwParseException(dateStr, df);
         }
-        return false;
+        
+        // Return result
+        //
+        return dt;
     }
-
-    private final static DateFormat _cloneFormat(DateFormat df, String format,
+    
+    protected void throwParseException(String dateStr, DateFormat df) throws ParseException
+    {
+    	if( df instanceof SimpleDateFormat ) {
+	    	 throw new ParseException
+		         (String.format("Can not parse date \"%s\": while it seems to fit format '%s', parsing fails (leniency? %s)",
+		                 dateStr, ((SimpleDateFormat)df).toPattern(), _lenient),
+		         0);
+    	}
+    	else {
+	    	 throw new ParseException
+		         (String.format("Can not parse date \"%s\" (leniency? %s)",
+		                 dateStr, _lenient),
+		         0);
+    	}
+    }
+    
+    protected int indexOfAnyFromEnd(String s, int startPos, int endPos, char... chars) 
```

Comment:
It looks like only 2 calls; one with 1 char (which can just use `String.lastIndexOf(ch)`), another with 2. So maybe write version for second case, to avoid inner loop as well as varargs.